File: sdk/src/math/bid.ts
/**
 * Deterministic agent bidding algorithm.
 *
 * ## Background
 *
 * When multiple agents compete for an escrow job the requester (or the
 * on-chain arbiter) needs a reproducible **bid score** so the same input
 * always produces the same winner, regardless of which CPU architecture
 * computed it.
 *
 * Historically this was done with native JS floats, which diverge between
 * x86 (SSE2) and ARM (M1/M2 NEON) due to intermediate-precision differences
 * in the FPU pipelines.  This file replaces every float operation with calls
 * to the `fixed-point` module, which uses `bignumber.js` pure-JS arithmetic.
 *
 * ## Scoring formula
 *
 * Each bid is scored 0–100 using a weighted combination of:
 *
 *   score = (priceWeight   × priceScore)
 *         + (repWeight     × repScore)
 *         + (latencyWeight × latencyScore)
 *         + (reliabWeight  × reliabScore)
 *
 * Where individual sub-scores are normalised to [0, 100]:
 *   - priceScore   = 100 × (1 − bid / maxBid)   — lower price is better
 *   - repScore     = reputation (0–100 input)
 *   - latencyScore = 100 × (1 − latency / maxLatency) — lower latency better
 *   - reliabScore  = successRate × 100           — 0.0–1.0 input
 *
 * All weights must sum to 1.0 (expressed as decimal strings).
 *
 * @module bid
 */

import BigNumber from 'bignumber.js';
import { bn, add, sub, mul, div, clamp } from './fixed-point.js';

// ─── Types ───────────────────────────────────────────────────────────────────

/** A single agent's bid for an escrow job */
export interface AgentBid {
  /** Unique agent address */
  agentAddress: string;
  /**
   * Price the agent is willing to accept for the job.
   * Must be a decimal string (e.g. "0.05") — never a JS number.
   */
  price: string;
  /**
   * Agent's reputation score 0–100 (integer or decimal string).
   * Sourced from on-chain historical data.
   */
  reputation: string;
  /**
   * Expected task completion time in seconds (decimal string).
   * Lower is better.
   */
  estimatedLatencySeconds: string;
  /**
   * Lifetime success rate as a decimal fraction 0–1 (e.g. "0.97").
   * Computed as successfulJobs / totalJobs.
   */
  successRate: string;
}

/** Weights controlling the relative importance of each dimension */
export interface BidWeights {
  /** Weight for price competitiveness (0–1, decimal string) */
  price: string;
  /** Weight for reputation (0–1, decimal string) */
  reputation: string;
  /** Weight for latency (0–1, decimal string) */
  latency: string;
  /** Weight for reliability / success rate (0–1, decimal string) */
  reliability: string;
}

/** Default weights — equal split across all four dimensions */
export const DEFAULT_BID_WEIGHTS: BidWeights = {
  price: '0.25',
  reputation: '0.25',
  latency: '0.25',
  reliability: '0.25',
};

/** A scored bid ready for ranking */
export interface ScoredBid {
  agentAddress: string;
  /** Final composite score 0–100, rounded down to 4 decimal places */
  score: string;
  /** Individual sub-scores for transparency */
  breakdown: {
    priceScore: string;
    reputationScore: string;
    latencyScore: string;
    reliabilityScore: string;
  };
}

// ─── Core scoring function ───────────────────────────────────────────────────

/**
 * Compute a deterministic composite score for a single bid.
 *
 * All intermediate values are `BigNumber` — no native JS floats are used.
 * The result is identical on x86, ARM, WASM, and any other platform where
 * `bignumber.js` runs.
 *
 * @param bid       - The agent's bid to score
 * @param maxBid    - The highest price among all competing bids (normaliser)
 * @param maxLatency - The highest latency among all competing bids (normaliser)
 * @param weights   - Dimension weights (must sum to 1)
 */
export function scoreBid(
  bid: AgentBid,
  maxBid: string,
  maxLatency: string,
  weights: BidWeights = DEFAULT_BID_WEIGHTS,
): ScoredBid {
  // ── Validate weights sum to 1.0 ──────────────────────────────────────────
  const weightSum = add(add(add(weights.price, weights.reputation), weights.latency), weights.reliability);
  if (!weightSum.isEqualTo(bn('1'))) {
    throw new RangeError(
      `BidScorer: weights must sum to 1.0, got ${weightSum.toFixed(4)}`,
    );
  }

  // ── Price score: 100 × (1 − price/maxBid) ────────────────────────────────
  // Lower price → higher score.  If all bids are equal maxBid === price → 0.
  let priceScore: BigNumber;
  if (bn(maxBid).isZero()) {
    priceScore = new BigNumber(100);
  } else {
    const priceRatio = div(bid.price, maxBid);        // price / maxBid ∈ [0,1]
    priceScore = mul(sub('1', priceRatio), '100');     // 100 × (1 − ratio)
    priceScore = clamp(priceScore, '0', '100');
  }

  // ── Reputation score: already 0–100 ──────────────────────────────────────
  const reputationScore = clamp(bn(bid.reputation), '0', '100');

  // ── Latency score: 100 × (1 − latency/maxLatency) ────────────────────────
  // Lower latency → higher score.
  let latencyScore: BigNumber;
  if (bn(maxLatency).isZero()) {
    latencyScore = new BigNumber(100);
  } else {
    const latencyRatio = div(bid.estimatedLatencySeconds, maxLatency);
    latencyScore = mul(sub('1', latencyRatio), '100');
    latencyScore = clamp(latencyScore, '0', '100');
  }

  // ── Reliability score: successRate × 100 ─────────────────────────────────
  const reliabilityScore = clamp(mul(bid.successRate, '100'), '0', '100');

  // ── Composite weighted score ──────────────────────────────────────────────
  const compositeScore = add(
    add(
      add(
        mul(priceScore, weights.price),
        mul(reputationScore, weights.reputation),
      ),
      mul(latencyScore, weights.latency),
    ),
    mul(reliabilityScore, weights.reliability),
  );

  return {
    agentAddress: bid.agentAddress,
    score: compositeScore.decimalPlaces(4, BigNumber.ROUND_DOWN).toFixed(4),
    breakdown: {
      priceScore: priceScore.decimalPlaces(4, BigNumber.ROUND_DOWN).toFixed(4),
      reputationScore: reputationScore.decimalPlaces(4, BigNumber.ROUND_DOWN).toFixed(4),
      latencyScore: latencyScore.decimalPlaces(4, BigNumber.ROUND_DOWN).toFixed(4),
      reliabilityScore: reliabilityScore.decimalPlaces(4, BigNumber.ROUND_DOWN).toFixed(4),
    },
  };
}

/**
 * Rank a set of competing agent bids deterministically.
 *
 * Steps:
 * 1. Compute `maxBid` and `maxLatency` over the full bid set (normalisation).
 * 2. Score every bid using `scoreBid`.
 * 3. Sort descending by score (ties broken by `agentAddress` lexicographically
 *    so the ordering is always reproducible).
 *
 * @returns Bids sorted best-first with their scores attached.
 */
export function rankBids(
  bids: AgentBid[],
  weights: BidWeights = DEFAULT_BID_WEIGHTS,
): ScoredBid[] {
  if (bids.length === 0) return [];

  // Normalisation pass — find max price and max latency across the whole set
  const maxBid = bids
    .map((b) => bn(b.price))
    .reduce((m, v) => (v.isGreaterThan(m) ? v : m))
    .toFixed(7);

  const maxLatency = bids
    .map((b) => bn(b.estimatedLatencySeconds))
    .reduce((m, v) => (v.isGreaterThan(m) ? v : m))
    .toFixed(7);

  // Score every bid
  const scored: ScoredBid[] = bids.map((bid) =>
    scoreBid(bid, maxBid, maxLatency, weights),
  );

  // Sort: highest score first; stable tie-break on address
  scored.sort((a, b) => {
    const cmp = bn(b.score).comparedTo(bn(a.score));
    if (cmp !== 0) return cmp;
    return a.agentAddress < b.agentAddress ? -1 : 1;
  });

  return scored;
}

/**
 * Select the single best bid from a pool.
 * Returns `null` when the pool is empty.
 */
export function selectBestBid(
  bids: AgentBid[],
  weights: BidWeights = DEFAULT_BID_WEIGHTS,
): ScoredBid | null {
  const ranked = rankBids(bids, weights);
  return ranked.length > 0 ? ranked[0]! : null;
}

// ─── Spend-limit helpers (reused from payment_channel contract logic) ─────────

/**
 * Determine whether a proposed payment is within an agent's spend limit.
 *
 * Replicates the on-chain `PaymentChannel.pay` guard in TypeScript so the SDK
 * can pre-validate without a network round-trip.  Uses the same integer-safe
 * arithmetic as the contract (amounts in stroops).
 *
 * @param spentThisPeriod  - Already spent in current period (stroop string)
 * @param limitPerPeriod   - Configured period limit (stroop string)
 * @param proposedAmount   - Amount about to be spent (stroop string)
 */
export function isWithinSpendLimit(
  spentThisPeriod: string,
  limitPerPeriod: string,
  proposedAmount: string,
): boolean {
  const newTotal = add(spentThisPeriod, proposedAmount);
  return newTotal.isLessThanOrEqualTo(bn(limitPerPeriod));
}

/**
 * Compute remaining budget in a period.
 *
 * @returns Remaining as a stroop decimal string (never negative).
 */
export function remainingBudget(
  spentThisPeriod: string,
  limitPerPeriod: string,
): string {
  const remaining = sub(limitPerPeriod, spentThisPeriod);
  return BigNumber.maximum(remaining, new BigNumber(0)).toFixed(0);
}

// ─── Re-export ────────────────────────────────────────────────────────────────
export { fmt, pct, add, sub, mul, div, toStroops, fromStroops } from './fixed-point.js';

File: python/src/stellaragent/bid.py
"""Deterministic agent bidding algorithm.

Python port of ``packages/core/src/math/bid.ts``. Every function here must
produce byte-identical strings to its TypeScript counterpart — see
``fixtures/determinism.json`` and the shared regression suite.

Scoring formula
---------------
Each bid is scored 0-100 as a weighted combination::

    score = (price_weight   * price_score)
          + (rep_weight     * reputation_score)
          + (latency_weight * latency_score)
          + (reliab_weight  * reliability_score)

with sub-scores normalised to [0, 100]:

- ``price_score``       = 100 * (1 - price / max_bid)      — lower is better
- ``reputation_score``  = reputation (already 0-100)
- ``latency_score``     = 100 * (1 - latency / max_latency) — lower is better
- ``reliability_score`` = success_rate * 100                — 0.0-1.0 input

Weights must sum to exactly 1.0. The check is exact equality rather than a
tolerance: an epsilon comparison would be sensitive to accumulated
representation error and defeat the whole point of the module.
"""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass, field
from decimal import Decimal

from .fixed_point import (
    FixedPointError,
    Numeric,
    _format,
    _quantize,
    add,
    bn,
    clamp,
    div,
    mul,
    sub,
)

__all__ = [
    "AgentBid",
    "BidWeights",
    "ScoredBid",
    "ScoreBreakdown",
    "DEFAULT_BID_WEIGHTS",
    "score_bid",
    "rank_bids",
    "select_best_bid",
    "is_within_spend_limit",
    "remaining_budget",
]


# ─── Types ───────────────────────────────────────────────────────────────────


@dataclass(frozen=True)
class AgentBid:
    """A single agent's bid for an escrow job.

    All numeric fields are decimal strings, never floats — the same rule the
    TS interface documents.
    """

    #: Unique agent address.
    agent_address: str
    #: Price the agent will accept, e.g. ``"0.05"``.
    price: str
    #: Reputation score 0-100, from on-chain history.
    reputation: str
    #: Expected completion time in seconds. Lower is better.
    estimated_latency_seconds: str
    #: Lifetime success rate as a fraction 0-1, e.g. ``"0.97"``.
    success_rate: str


@dataclass(frozen=True)
class BidWeights:
    """Relative importance of each scoring dimension. Must sum to exactly 1."""

    price: str = "0.25"
    reputation: str = "0.25"
    latency: str = "0.25"
    reliability: str = "0.25"


@dataclass(frozen=True)
class ScoreBreakdown:
    """Individual sub-scores, for transparency. Each is a 4-decimal string."""

    price_score: str
    reputation_score: str
    latency_score: str
    reliability_score: str


@dataclass(frozen=True)
class ScoredBid:
    """A scored bid ready for ranking."""

    agent_address: str
    #: Final composite score 0-100, truncated to 4 decimal places.
    score: str
    breakdown: ScoreBreakdown = field(compare=True)


#: Equal split across all four dimensions.
DEFAULT_BID_WEIGHTS = BidWeights()


# ─── Core scoring ────────────────────────────────────────────────────────────


def score_bid(
    bid: AgentBid,
    max_bid: Numeric,
    max_latency: Numeric,
    weights: BidWeights = DEFAULT_BID_WEIGHTS,
) -> ScoredBid:
    """Compute a deterministic composite score for one bid.

    :param bid: the bid to score
    :param max_bid: highest price among competing bids (the normaliser)
    :param max_latency: highest latency among competing bids
    :param weights: dimension weights, which must sum to 1
    :raises FixedPointError: when the weights do not sum to exactly 1
    """
    weight_sum = add(add(add(weights.price, weights.reputation), weights.latency), weights.reliability)
    if weight_sum != bn("1"):
        raise FixedPointError(
            f"BidScorer: weights must sum to 1.0, got {_format(_quantize(weight_sum, 4))}"
        )

    # Price: 100 * (1 - price/max_bid). A zero normaliser means every bid is
    # free, so short-circuit to full marks rather than dividing by zero.
    if bn(max_bid).is_zero():
        price_score = Decimal(100)
    else:
        price_ratio = div(bid.price, max_bid)
        price_score = clamp(mul(sub("1", price_ratio), "100"), "0", "100")

    reputation_score = clamp(bn(bid.reputation), "0", "100")

    # Latency: 100 * (1 - latency/max_latency). Same zero-normaliser case.
    if bn(max_latency).is_zero():
        latency_score = Decimal(100)
    else:
        latency_ratio = div(bid.estimated_latency_seconds, max_latency)
        latency_score = clamp(mul(sub("1", latency_ratio), "100"), "0", "100")

    reliability_score = clamp(mul(bid.success_rate, "100"), "0", "100")

    composite = add(
        add(
            add(
                mul(price_score, weights.price),
                mul(reputation_score, weights.reputation),
            ),
            mul(latency_score, weights.latency),
        ),
        mul(reliability_score, weights.reliability),
    )

    return ScoredBid(
        agent_address=bid.agent_address,
        score=_fixed4(composite),
        breakdown=ScoreBreakdown(
            price_score=_fixed4(price_score),
            reputation_score=_fixed4(reputation_score),
            latency_score=_fixed4(latency_score),
            reliability_score=_fixed4(reliability_score),
        ),
    )


def rank_bids(
    bids: Sequence[AgentBid],
    weights: BidWeights = DEFAULT_BID_WEIGHTS,
) -> list[ScoredBid]:
    """Rank competing bids deterministically, best first.

    Normalises against the maximum price and latency across the whole set,
    scores every bid, then sorts descending by score with ties broken
    lexicographically on ``agent_address`` so the ordering is reproducible
    regardless of input order.
    """
    if not bids:
        return []

    # The normalisers are truncated to 7 decimal places *before* scoring, as
    # in the TS original — this is observable in the final digit.
    max_bid = _fixed7(max(bn(b.price) for b in bids))
    max_latency = _fixed7(max(bn(b.estimated_latency_seconds) for b in bids))

    scored = [score_bid(bid, max_bid, max_latency, weights) for bid in bids]

    # Sort by descending score, then ascending address. Python's sort is
    # stable and total, so this needs no explicit comparator.
    scored.sort(key=lambda s: (-bn(s.score), s.agent_address))
    return scored


def select_best_bid(
    bids: Sequence[AgentBid],
    weights: BidWeights = DEFAULT_BID_WEIGHTS,
) -> ScoredBid | None:
    """Select the single best bid, or ``None`` when the pool is empty."""
    ranked = rank_bids(bids, weights)
    return ranked[0] if ranked else None


# ─── Spend limits ────────────────────────────────────────────────────────────


def is_within_spend_limit(
    spent_this_period: Numeric,
    limit_per_period: Numeric,
    proposed_amount: Numeric,
) -> bool:
    """Whether a proposed payment stays within the spend limit.

    Replicates the on-chain ``PaymentChannel.pay`` guard so the SDK can
    pre-validate without a network round-trip. Inclusive at the limit.
    """
    return add(spent_this_period, proposed_amount) <= bn(limit_per_period)


def remaining_budget(spent_this_period: Numeric, limit_per_period: Numeric) -> str:
    """Remaining budget as an integer stroop string, never negative."""
    remaining = sub(limit_per_period, spent_this_period)
    return _format(_quantize(max(remaining, Decimal(0)), 0))


# ─── Internals ───────────────────────────────────────────────────────────────


def _fixed4(value: Decimal) -> str:
    """The TS ``.decimalPlaces(4, ROUND_DOWN).toFixed(4)``."""
    return _format(_quantize(value, 4))


def _fixed7(value: Decimal) -> str:
    """The TS ``.toFixed(7)`` applied to the normalisers."""
    return _format(_quantize(value, 7))
